File: rotary_encoder.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon May 10 16:40:57 2021

@author: nivethithakannan
"""
from RPi import GPIO
from time import sleep

PIN_CLK = 17
PIN_DAT = 18

GPIO.setmode(GPIO.BCM)

try:
    GPIO.setup(PIN_CLK,GPIO.IN)
    GPIO.setup(PIN_DAT,GPIO.IN)
    GPIO.output(PIN_CLK,1)
except:
    print("ERROR. Unable to setup the configuration requested"  )                                   


sleep(0.5)
counter = 0
clkLastState = GPIO.input(PIN_CLK)

try:

        while True:
                clkState = GPIO.input(PIN_CLK)
                dtState = GPIO.input(PIN_DAT)
                if clkState != clkLastState:
                        if dtState != clkState:
                                counter += 1
                        else:
                                counter -= 1
                        print(counter)
                clkLastState = clkState
                sleep(0.01)
finally:
        GPIO.cleanup()

File: rotary_encoder.rb
require 'rpi_gpio'

RPi::GPIO.set_numbering :bcm

CLOCK = 17
DATA = 18

RPi::GPIO.setup(CLOCK, as: :input)
RPi::GPIO.setup(DATA, as: :input)

def dtState
  RPi::GPIO.high?(DATA) ? 1 : 0
end

def clock_state
  RPi::GPIO.high?(CLOCK) ? 1 : 0
end

counter = 0
clock_last_state = clock_state

loop do
  if clock_state != clock_last_state
    if dtState != clock_state
      counter += 1
    else
      counter -= 1
    end

    puts counter
  end

  clock_last_state = clock_state
  sleep 0.01
end

at_exit do
  RPi::GPIO.clean_up
end
